sklearn/cluster/_k_means.pyx

```diff
@@ -8,161 +8,257 @@
-            center_squared_norms[center_idx] = _dot(
-                n_features, &centers[center_idx, 0], 1,
-                &centers[center_idx, 0], 1)
-
-    for sample_idx in range(n_samples):
-        min_dist = -1
-        for center_idx in range(n_clusters):
-            dist = 0.0
-            # hardcoded: minimize euclidean distance to cluster center:
-            # ||a - b||^2 = ||a||^2 + ||b||^2 -2 <a, b>
-            for k in range(X_indptr[sample_idx], X_indptr[sample_idx + 1]):
-                dist += centers[center_idx, X_indices[k]] * X_data[k]
-            dist *= -2
-            dist += center_squared_norms[center_idx]
-            dist += x_squared_norms[sample_idx]
-            dist *= sample_weight[sample_idx]
-            if min_dist == -1 or dist < min_dist:
-                min_dist = dist
-                labels[sample_idx] = center_idx
-                if store_distances:
-                    distances[sample_idx] = dist
-        inertia += min_dist
+    for i in range(n_samples):
+        j = labels[i]
+        sq_dist = _euclidean_sparse_dense(
+            X_data[X_indptr[i]: X_indptr[i + 1]],
+            X_indices[X_indptr[i]: X_indptr[i + 1]],
+            centers[j], centers_squared_norms[j], True)
+        inertia += sq_dist * sample_weight[i]
 
     return inertia
 
 
-@cython.boundscheck(False)
-@cython.wraparound(False)
-@cython.cdivision(True)
+cdef void _relocate_empty_clusters_dense(np.ndarray[floating, ndim=2, mode='c'] X,
+                                         floating[::1] sample_weight,
+                                         floating[:, ::1] centers_old,
+                                         floating[:, ::1] centers_new,
+                                         floating[::1] weight_in_clusters,
+                                         int[::1] labels):
+    """Relocate centers which have no sample assigned to them."""
+    cdef:
+        int[::1] empty_clusters = np.where(np.equal(weight_in_clusters, 0))[0].astype(np.int32)
+        int n_empty = empty_clusters.shape[0]
+
+    if n_empty == 0:
+        return
+
+    cdef:
+        int n_features = X.shape[1]
+
+        floating[::1] distances = ((np.asarray(X) - np.asarray(centers_old)[labels])**2).sum(axis=1)
+        int[::1] far_from_centers = np.argpartition(distances, -n_empty)[:-n_empty-1:-1].astype(np.int32)
+
+        int new_cluster_id, old_cluster_id, far_idx, idx, k
+        floating weight
+
+    for idx in range(n_empty):
+
+        new_cluster_id = empty_clusters[idx]
+
+        far_idx = far_from_centers[idx]
+        weight = sample_weight[far_idx]
+
+        old_cluster_id = labels[far_idx]
+
+        for k in range(n_features):
+            centers_new[old_cluster_id, k] -= X[far_idx, k] * weight
+            centers_new[new_cluster_id, k] = X[far_idx, k] * weight
+
+        weight_in_clusters[new_cluster_id] = weight
+        weight_in_clusters[old_cluster_id] -= weight
+    print('ok')
```

Comment:
Happy to see we share the same advanced debugging techniques ;)

Reply:
I struggled with a segfault for hours :D